### ISSUE TITLE
BCEwithlogitsloss for 18 classes

### ISSUE THREAD

**Issue by nikhilchh** [wmcnally/kapao#42]
The question is why do yolo (and also Kapao) uses BCEwithlogitsLoss for 18 classes?
1- Yolo and Kapao is multi class single label problem.
2- But the above loss is good for multi class multi label. As it doesn't uses Soft-max

**Comment by wmcnally** [comment]
With yolov5 (and maybe previous versions?) the multi-class problem is reformulated as binary classification. I agree it's an unusual way to apply a multi-class loss. Here's how I understand it:

**A binary cross entropy (BCE) loss is applied at every cell in the output grids. There are two BCE losses: one for if an object exists in that cell, and one for the class of object in that cell.** 

The object / no object is a standard BCE loss (1: object, 0: no object) and is applied at every grid cell. 

The BCE loss for the object classes works similarly. Imagine each class having its own set of output grids with 1s in cells where an object of that class exists and 0s elsewhere. The class BCE loss is applied at every cell in those class grids. 

Hope this helps.

**Comment by nikhilchh** [comment]
Thanks for your response.

Questions:
1- Why are they doing this instead of going for softmax ? In theory do both approaches leads to similar results ?

The more important question for me is:
How are you making decision at the end in the code ? During inference time do you apply sigmoid to each of the 18 values and then normalize them and choose the one with max value ?

**Comment by wmcnally** [comment]
FYI, sigmoid is applied inside the [BCEwithlogitsLoss](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html) function. Still, I'm not sure exactly why the class loss is reformulated this way. I think with sufficient loss weight tuning, both approaches (binary vs. multi-class) should be able to achieve similar results. 

At inference time, sigmoid is applied to the logits so each grid cell has class scores ranging from 0 to 1. The class with the highest score is chosen.
